akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala

```diff
@@ -116,168 +130,217 @@ private class ResumeSupervisor[T, Thr <: Throwable: ClassTag](ss: Resume) extend
   }
 
-  protected def handleExceptionOnStart(ctx: ActorContext[O]): Catcher[Behavior[T]] = {
-    case NonFatal(t: Thr) ⇒
-      scheduleRestart(ctx, t)
-  }
+  private def prepareRestart(ctx: ActorContext[O], reason: Throwable): Behavior[T] = {
+    log(ctx, reason)
 
-  protected def handleReceiveException(ctx: ActorContext[O], target: ReceiveTarget[T]): Catcher[Behavior[T]] = {
-    case NonFatal(t: Thr) ⇒
-      try {
-        target.signalRestart(ctx)
-      } catch {
-        case NonFatal(ex) ⇒ ctx.asScala.log.error(ex, "failure during PreRestart")
-      }
-      scheduleRestart(ctx, t)
+    val currentRestartCount = restartCount
+    increaseRestartCount()
+
+    val childrenToStop = if (strategy.stopChildren) ctx.asScala.children.toSet else Set.empty[ActorRef[Nothing]]
+    stopChildren(ctx, childrenToStop)
+
+    val stashCapacity =
+      if (strategy.stashCapacity >= 0) strategy.stashCapacity
+      else ctx.asScala.system.settings.RestartStashCapacity
+    restartingInProgress = OptionVal.Some((StashBuffer[Any](stashCapacity), childrenToStop))
+
+    strategy match {
+      case backoff: Backoff ⇒
+        val restartDelay = calculateDelay(currentRestartCount, backoff.minBackoff, backoff.maxBackoff, backoff.randomFactor)
+        gotScheduledRestart = false
+        ctx.asScala.scheduleOnce(restartDelay, ctx.asScala.self.unsafeUpcast[Any], ScheduledRestart)
+        Behaviors.empty
+      case _: Restart ⇒
+        if (childrenToStop.isEmpty)
+          restartCompleted(ctx)
+        else
+          Behaviors.empty // wait for termination of children
+    }
   }
 
-  protected def handleSignalException(ctx: ActorContext[O], target: SignalTarget[T]): Catcher[Behavior[T]] = {
-    case NonFatal(t: Thr) ⇒
-      try {
-        target(ctx, PreRestart)
-      } catch {
-        case NonFatal(ex) ⇒ ctx.asScala.log.error(ex, "failure during PreRestart")
+  private def restartCompleted(ctx: ActorContext[O]): Behavior[T] = {
+    strategy match {
+      case backoff: Backoff ⇒
+        gotScheduledRestart = false
+        ctx.asScala.scheduleOnce(backoff.resetBackoffAfter, ctx.asScala.self.unsafeUpcast[Any], ResetRestartCount(restartCount))
+      case _: Restart ⇒
+    }
+
+    try {
+      val newBehavior = Behavior.validateAsInitial(Behavior.start(initial, ctx.asInstanceOf[ActorContext[T]]))
+      val nextBehavior = restartingInProgress match {
+        case OptionVal.None ⇒ newBehavior
+        case OptionVal.Some((stashBuffer, _)) ⇒
+          restartingInProgress = OptionVal.None
+          stashBuffer.unstashAll(ctx.asScala.asInstanceOf[scaladsl.ActorContext[Any]], newBehavior.unsafeCast)
       }
-      scheduleRestart(ctx, t)
+      nextBehavior.narrow
+    } catch handleException(ctx, signalRestart = () ⇒ ())
+    // FIXME signal Restart is not done if unstashAll throws, unstash of each message may return a new behavior and
+    //      it's the failing one that should receive the signal
```

Comment:
will not do it in this PR but added issue https://github.com/akka/akka/issues/26148

Reply:
I will start #26148 after this is merged.

Reply:
@helena I merged this